lang/src/org/partiql/lang/eval/physical/operators/AggregateOperatorFactory.kt

```diff
@@ -0,0 +1,172 @@
+                    )
+                }
+
+                // Aggregate Values in Aggregation State
+                functions.forEachIndexed { index, function ->
+                    val valueToAggregate = function.value(state)
+                    when (function.quantifier) {
+                        is PartiqlPhysical.SetQuantifier.Distinct -> {
+                            if (seen[index].add(valueToAggregate)) {
+                                function.function(groupMapValue.state, valueToAggregate)
+                            }
+                        }
+                        is PartiqlPhysical.SetQuantifier.All -> {
+                            function.function(groupMapValue.state, valueToAggregate)
+                        }
+                    }
+                }
+            }
+
+            // Place Aggregated Values into Result State
+            relation(sourceIter.relType) {
+                if (keys.isEmpty() && aggregationMap.isEmpty()) {
+                    functions.forEach { function ->
+                        function.function(state, state.valueFactory.nullValue)
+                    }
+                    yield()
+                    return@relation
+                }
+                aggregationMap.forEach { (ionValues, groupMapValue) ->
+                    transferRegisters(groupMapValue.state, state, variablesIndexes)
+                    ionValues.forEachIndexed { index, ionValue ->
+                        if (ionValue.isNullValue && groupMapValue.isMissingArray[index]) {
+                            ionValue.addTypeAnnotation(MISSING_ANNOTATION)
+                        }
+                        keys[index].function.invoke(state, state.valueFactory.newFromIonValue(ionValue))
+                    }
+                    yield()
+                }
+            }
+        }
+}
+
+/**
+ * @param state holds the aggregated values in its registers
+ * @param isMissingArray represents whether any found ion NULL values are actually MISSING. This needs to be represented
+ *  externally so that the equality comparison for the keys in the [AggregationMap] can be stored together. See
+ *  https://github.com/partiql/partiql-lang-kotlin/issues/35 for context.
+ */
+private class GroupMapValue(val state: EvaluatorState, val isMissingArray: Array<Boolean>)
+
+/**
+ * A [LinkedHashMap] indexed by a [List] of [IonValue]s which implement [equals] and [hashCode]. This allows us to index
+ * into the map after evaluating the [CompiledGroupKey]. Then, the [GroupMapValue] is a representation of the total
+ * aggregations, and it holds metadata to help differentiate between [ExprValueType.MISSING] and [ExprValueType.NULL]
+ */
+private typealias AggregationMap = LinkedHashMap<List<IonValue>, GroupMapValue>
+
+public class CompiledGroupKey(
+    val function: SetVariableFunc,
+    val value: ValueExpression,
+    val variable: PartiqlPhysical.VarDecl
+)
+
+public class CompiledAggregateFunction(
+    val function: SetVariableFunc,
```

Comment:
```suggestion
    val setAggregateVal: SetVariableFunc,
```

function.function is quite difficult to read.